coap/adapter.go

```diff
@@ -135,10 +138,27 @@ func (svc *adapterService) Publish(ctx context.Context, token string, msg mainfl
 }
 
 func (svc *adapterService) Subscribe(chanID, subtopic, obsID string, o *Observer) error {
-	if err := svc.pubsub.Subscribe(chanID, subtopic, obsID, o); err != nil {
-		return ErrFailedSubscription
+	sub, err := svc.pubsub.Subscribe(chanID, subtopic, func(msg *nats.Msg) {
+		if msg == nil {
+			return
+		}
+		var m broker.Message
+		if err := proto.Unmarshal(msg.Data, &m); err != nil {
+			return
+		}
+		o.Messages <- m
+	})
+	if err != nil {
+		return err
 	}
 
+	go func() {
+		<-o.Cancel
```

Comment:
I may be wrong but seems that wait group fits here, us if you can, it is faster.